packages/dds/tree/src/codec/versioned/codec.ts

```diff
@@ -140,30 +164,320 @@ export function makeVersionDispatchingCodec<TDecoded, TContext>(
+		return {
+			name: this.name,
+			version: selected.formatVersion,
+		};
 	}
+
+	/**
+	 * Builds a ClientVersionDispatchingCodecBuilder from the provided registry.
+	 * @remarks
+	 * This static method infers the types of the builder from the provided registry,
+	 * making it easier to create builders without needing to explicitly specify all type parameters.
+	 * This gets better type inference than the constructor.
+	 */
+	// eslint-disable-next-line @typescript-eslint/explicit-function-return-type
+	public static build<
+		Name extends CodecName,
+		Entry extends CodecVersion<unknown, unknown, FormatVersion, never>,
+	>(name: Name, inputRegistry: ConfigMapEntry<Entry>) {
+		type TDecoded2 = Entry extends CodecVersion<infer D, unknown, FormatVersion, never>
+			? D
+			: never;
+		type TContext2 = Entry extends CodecVersion<unknown, infer C, FormatVersion, never>
+			? C
+			: never;
+		type TFormatVersion2 = Entry extends CodecVersion<unknown, unknown, infer F, never>
+			? F
+			: never;
+		type TBuildOptions2 = Entry extends CodecVersion<unknown, unknown, FormatVersion, infer B>
+			? B
+			: never;
+		const builder = new ClientVersionDispatchingCodecBuilder(
+			name,
+			inputRegistry as ConfigMapEntry<unknown> as ConfigMapEntry<
+				CodecVersion<
+					TDecoded2,
+					// If it does not matter what context is provided, undefined is fine, so allow it to be omitted.
+					TContext2 extends unknown ? void : TContext2,
+					TFormatVersion2,
+					TBuildOptions2
+				>
+			>,
+		);
+		return builder;
+	}
+}
+
+/**
+ * Selects which format should be used when writing data.
+ * @remarks
+ * This either uses the override specified in the options, or selects the newest format compatible with the provided minVersionForCollab.
+ */
+function getWriteVersion<T extends CodecVersionBase>(
+	name: CodecName,
+	options: CodecWriteOptions,
+	versions: readonly [MinimumMinorSemanticVersion | MinimumVersionForCollab, T][],
+): T {
+	if (options.writeVersionOverrides?.has(name) === true) {
+		const selectedFormatVersion = options.writeVersionOverrides.get(name);
+		const selected = versions.find(
+			([_v, codec]) => codec.formatVersion === selectedFormatVersion,
+		);
+		if (selected === undefined) {
+			throw new UsageError(
+				`Codec "${name}" does not support requested format version ${selectedFormatVersion}. Supported versions are: ${versionList(versions)}.`,
+			);
+		} else if (options.allowPossiblyIncompatibleWriteVersionOverrides !== true) {
+			const selectedMinVersionForCollab = selected[0];
+			if (selectedMinVersionForCollab === undefined) {
+				throw new UsageError(
+					`Codec "${name}" does not support requested format version ${selectedFormatVersion} because it does not specify an oldest compatible client. Use "allowPossiblyIncompatibleWriteVersionOverrides" to override this error.`,
+				);
+			} else if (gt(selectedMinVersionForCollab, options.minVersionForCollab)) {
+				throw new UsageError(
```

Comment:
See above comment about potentially misleading the reader into thinking that they should for sure do that when you encounter this error.

Reply:
Also fixed.